service/src/test/java/bio/terra/catalog/api/DatasetApiControllerTest.java

```diff
@@ -74,6 +74,20 @@ void listDatasets() throws Exception {
         .andExpect(jsonPath("$.result[0].id").value("id"));
   }
 
+  @Test
+  void EmptyListDatasets() throws Exception {
+    DatasetsListResponse response = new DatasetsListResponse();
+    ObjectNode node = new ObjectMapper().createObjectNode();
+    response.addResultItem(node);
+    when(datasetService.listDatasets(user)).thenReturn(response);
```

Comment:
Because this datasetService call is mocked, I am not sure that this test adds much.

I would probably rather see a test that calls the underlying method (maybe with a mock on `datasetDAO` to ensure we get no values back from the database?) and at the end something like:
`assertThat(response.getResults().getSize()).isZero()`

Reply:
I think this is the correct way to test this. The purpose of this test is to see if the JSON result is as expected or not. This does require that we create a response object that correctly matches what we'd get if there are no datasets to return, and I think removing the call to `response.addResultItem()` is required to do that.

Regarding `assertThat(response.getResults().getSize()).isZero()`, I think `.andExpect(jsonPath("$.result[0]").isEmpty())` fulfills this requirement (maybe your comment refers to an earlier version of this test?).

> I would probably rather see a test that calls the underlying method (maybe with a mock on datasetDAO to ensure we get no values back from the database?

If this test were to mock at the DAO layer, then it would have to use a "real" `DatasetService` instead of a mocked one, which would change the way all test methods in this class works. In general I think we want a test to mock everything except the class under test itself. (Except for DAO tests, which require a live database.)

Reply:
In my mind, the place to test is at the place where if it found no items, it generated the response object. My concern here is that we're mocking out the actual response object, which means we aren't testing our code, we're testing what swagger is doing. Given that the change was just done in swagger, I actually am not sure we should test this anymore.

If we do need to test it, I would probably do:
```
DatasetsListResponse response = new DatasetsListResponse();
assertThat(response.getResultItems().size(), assertDoesNotThrow())
```

Because that is actually the behavior that we want to check here (that doing operations on the result set doesn't fail even if it is empty)